File: kotlin-runtime/src/main/kotlin/rosettarpc/RpcErrorCode.kt
package rosettarpc

/**
 * Stable, wire-level discriminant for an [RpcException]. Must stay in sync across all language
 * runtimes.
 */
enum class RpcErrorCode(val wireValue: Int) {
  UNKNOWN_SERVICE(0),
  UNKNOWN_METHOD(1),
  DECODE(2),

  /**
   * The service implementation itself failed: either it deliberately threw, or an otherwise-
   * uncaught exception was caught on its behalf by `GodotSupport.dispatchBytes`.
   */
  APPLICATION(3);

  companion object {
    fun fromWireValue(value: Int): RpcErrorCode =
        entries.find { it.wireValue == value } ?: APPLICATION
  }
}

File: csharp-runtime/RpcErrorCode.cs
namespace RosettaRpc;

/// <summary>
/// Wire-level discriminant for an <see cref="RpcException"/>. Must
/// stay in sync across all language runtimes.
/// </summary>
public enum RpcErrorCode
{
    UnknownService = 0,
    UnknownMethod = 1,
    Decode = 2,

    /// <summary>
    /// The service implementation itself failed: either it deliberately
    /// threw, or an otherwise-uncaught exception was caught on its behalf by
    /// <c>GodotSupport.DispatchBytes</c>.
    /// </summary>
    Application = 3,
}
